google/cloud/pubsub/internal/tracing_message_batch.h

```diff
@@ -0,0 +1,64 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_PUBSUB_INTERNAL_TRACING_MESSAGE_BATCH_H
+#define GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_PUBSUB_INTERNAL_TRACING_MESSAGE_BATCH_H
+
+#ifdef GOOGLE_CLOUD_CPP_HAVE_OPENTELEMETRY
+
+#include "google/cloud/pubsub/internal/message_batch.h"
+#include "google/cloud/pubsub/internal/publisher_stub.h"
+#include "google/cloud/pubsub/version.h"
+#include "google/cloud/internal/opentelemetry.h"
+#include <memory>
+
+namespace google {
+namespace cloud {
+namespace pubsub_internal {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+
+class TracingMessageBatchPeer;
+
+/**
+ * Records spans related to a batch messages across calls and
+ * callbacks in the `BatchingPublisherConnection`.
+ */
+class TracingMessageBatch : public MessageBatch {
+ public:
+  explicit TracingMessageBatch(std::unique_ptr<MessageBatch> child)
+      : child_(std::move(child)) {}
+  ~TracingMessageBatch() override = default;
+
+  void SaveMessage(pubsub::Message m) override;
+
+  void Flush() override;
+
+  void FlushCallback() override;
+
+ private:
+  std::unique_ptr<MessageBatch> child_;
+  std::vector<opentelemetry::nostd::shared_ptr<opentelemetry::trace::Span>>
+      message_root_spans_;
```

Comment:
Maybe just `message_spans`. This could be non-root spans if they happen in the context of some other operation?